### PR TITLE
12094 visualization invalidation

### LINKED ISSUE #12094: [Model refactor] Named maps update/Cache invalidation
STOP THIS MADNESS!

### PR DESCRIPTION
**Pull request by javitonino**
Closes #12094 

**Acceptance**

This PR is totally related with caches, so please check request headers all the time :-)

Some things to test, feel free to add:
- [x] Don't forget about editor! It's probably more important than builder for this.
- [x] Mapcap interactions
  - [x] Privacy/permissions should always be updated (even without publishing). Best way to try is to switch from private to public. If the named map update failed, you will not see anything (missing auth tokens)
  - [x] Other changes should only be made public on map publication
- [x] Update different elements and see that caches are invalidated, named maps published. Try without a mapcap, because mapcapping explicitly invalidates everything.
  - [x] Visualization (privacy, metadata)
  - [x] Map (state, map options)
  - [x] Layers
  - [x] Widgets
  - [x] Analyses
  - [x] Overlays
- [x] Changing dataset metadata should invalidate embeds of map using it (see in the public view, the list of used dataset is displayed, including the description)
- [x] Map destruction (you can use named maps api index to see it has been destroyed, cache is trickier)
- [x] Dataset destruction (should invalidate affected maps)

---

The purpose of this PR is to move `notify_map_change` so it does not use the old models at all.

Summary:
- Extracted invalidation logic to a service
- Cleaned up a lot of decisions in the invalidation. It should be simpler now, and extracted in a separate service.
- Changed the flow
Before this PR, flow is:
1. `Carto::Map.notify_map_change`
2. `::Map.notify_map_change`
3. `::Map.update_related_named_maps`
4. `Visualization::Member.save_named_map`
5. `Rails::Sequel.connection.after_commit`
6. `Visualization::Member.update_named_map`
7. `Carto::NamedMaps::Api.update`

After this PR
1. `Carto::Map.notify_map_change`
2. `Carto::Visualization.invalidate_later`
3. `Carto::Visualization.after_commit` hook
4. `Carto::Visualization.perform_invalidations`
5. `Carto::VisualizationInvalidationService.invalidate`

**Deployment**

If something's wrong with this PR it might increase the load (if caching doesn't work properly) or stop updating maps (if invalidation is wrong), so ping #systems and #support, and monitor the load.

### CODE REVIEW

**Comment by javitonino** [comment]
This is starting to take shape, before going on, I'd like a quick discussion. See the description for some context, which I'll complete below.

There are two changes in this PR:
- Move the named map updates so the flow goes from `Carto::Map` to `Carto::Visualization` without passing through the old models.
- The named map update is called outside of the transaction for new models (we were already doing this in the old models)

The latter uses `add_to_transaction`, which is a documented AR function, although a bit obscure. Basically, this add the model to the list of models touched during the transaction, so its `after_commit` hook is called. This is not too bad, but also not great.

Why are we doing this? Basically to avoid running named maps update in `after_save` (which runs inside the transaction). Note that right now, cache invalidations do run inside the transaction (and some of those imply a `telnet` connection to Varnish).

The question is about how far do we want to do with this refactor. Some ideas:
1. Keep the current code, which keeps the current behaviour as closely as possible, just using AR instead of the old ORMs.
2. Move cache invalidation to the `after_commit` hook as well, moving it out of the transaction (as a bonus, it only runs once per transaction).
3. Move to using TIS instead of direct Varnish connection (this is a [separate issue](https://github.com/CartoDB/cartodb/issues/11513) but I wanted to mention it)
4. Move this to a service

I am thinking that this is a chance to start moving some stuff. I would like to approach 2 for sure. This changes a bit the behaviour, since right now, we are not invalidating the cache for every visualization change (only some privacy related ones). But, as we are invalidating anyway each time we change the map, its layers, etc. this micro-optimization seems pointless and overly complicated.

The good part about moving this change, is that we can have a single method that invalidates everything related to the visualization (we could even move it out of `Carto::Visualization` as the skeleton of an invalidation service). For now, we would call it from the `after_commit` hook of visualization, and we will keep the current callbacks to ensure this `after_commit` hook is called when touching any model related to the visualization.

We can go much further, but I think this is a pretty good step in the good direction, which fits within the scope of this PR.

@CartoDB/backend

**Comment by juanignaciosl** [comment]
Agree on trying approach 2.

File: Makefile [hunk 1]
```diff
@@ -301,6 +301,7 @@ SPEC_HELPER_MIN_SPECS = \
 	spec/lib/carto/assets_service_spec.rb \
 	spec/lib/carto/organization_assets_service_spec.rb \
 	spec/lib/carto/storage_options/local_spec.rb \
+	spec/lib/carto/visualization_invalidation_service.rb \
```
Comment (juanignaciosl) [inline Makefile]: ❤️

File: app/models/carto/layer.rb [hunk 2]
```diff
@@ -201,7 +201,7 @@ def supports_labels_layer?
     end
 
     def map
-      maps.first
+      maps[0]
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: What's the reason for this change? (sincere question)

File: app/models/carto/visualization.rb [hunk 3]
```diff
@@ -92,20 +91,20 @@ class Carto::Visualization < ActiveRecord::Base
...
-  after_destroy :invalidate_cache
-  after_destroy :destroy_named_map
 
   # INFO: workaround for array saves not working. There is a bug in `activerecord-postgresql-array` which
   # makes inserting including array fields to save, but updates work. Wo se insert without tags and add them
   # with an update after creation. This is fixed in Rails 4.
   before_create :delay_saving_tags
   after_create :save_tags
 
+  after_commit :perform_invalidations
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Ok, we could say that this is the core of the change. This works only if there's no change operation on old models (visualization, map, layer), isn't it?

File: app/models/carto/visualization.rb [hunk 4]
```diff
@@ -556,29 +533,31 @@ def unlink_from(user_table)
     end
   end
 
+  def invalidate_later
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: - Minor: I'd explicitly name this after the implementation: `invalidate_after_commit`, for example, because in this case this is quite specific and with many implications. Also, you can move the comment on top of the method.

- Would this implementation workaround the fact of having different transactions?

- If current implementation only invalidated once within the transaction and the invalidation failed the transaction would end in a rollback. Nevertheless, with this change, the transaction would commit and then the invalidation might fail, leading to some inconsistencies. Am I right? Previous implementation can never lead to _this_ inconsistency, isn't it? I don't think that this is a severe concern, but I want to make sure that I get the consequences of this change and that we're aware of every possible side effect.

File: app/models/carto/visualization.rb [hunk 5]
```diff
@@ -556,29 +533,31 @@ def unlink_from(user_table)
...
+    # This marks this visualization as affected by this transaction, so AR will call its `after_commit` hook, which
+    # performs the actual invalidations. This takes this operation outside of the DB transaction to avoid long locks
+    add_to_transaction
+    true
+  end
+  # TODO: Privacy manager compatibility, can be removed after removing ::usUserTable
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: `usUserTable` typo.

File: lib/carto/visualization_invalidation_service.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      @invalidate_affected_visualizations = false
+    end
+
+    def invalidate
+      invalidate_caches
+      update_or_destroy_named_map
+      invalidate_affected_visualizations if @invalidate_affected_visualizations
+    end
+
+    def with_invalidation_of_affected_visualizations
```
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: Minor: `with_` notation is confusing here. Services are not the kind of objects that need fluent interfaces, and naming is worse. Consider `enable_invalidation_of_affected_visualizations`, for example.

File: lib/carto/visualization_invalidation_service.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
+module Carto
+  class VisualizationInvalidationService
+    def initialize(visualization)
+      @visualization = visualization
+      @invalidate_affected_visualizations = false
+    end
+
+    def invalidate
+      invalidate_caches
```
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: Probably minor: shouldn't cache invalidation be the last step?

File: lib/carto/visualization_invalidation_service.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    private
+
+    def update_or_destroy_named_map
+      return if @visualization.remote?
+      named_maps_api = Carto::NamedMaps::Api.new(@visualization.for_presentation)
+      if @visualization.destroyed?
+        named_maps_api.destroy
+      elsif @visualization.data_layers.any?
+        named_maps_api.show ? named_maps_api.update : named_maps_api.create
```
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: I'd add an `upsert` method at `Carto::NamedMaps::Api`.

File: lib/carto/visualization_invalidation_service.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      end
+    end
+
+    def invalidate_affected_visualizations
+      @visualization.user_table.dependent_visualizations.each do |affected_visualization|
+        VisualizationInvalidationService.new(affected_visualization).invalidate_caches
```
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: Shouldn't this `invalidate`, instead of `invalidate_caches`?

File: app/models/carto/visualization.rb [hunk 10]
```diff
@@ -556,29 +533,31 @@ def unlink_from(user_table)
...
-    # now we need to purgue everything to avoid cached stale data or public->priv still showing scenarios
-    if privacy_changed? || name_changed? || cached_data_changed?
-      invalidate_cache
-    end
-
+  def invalidate_related_visualizations
     # When a table's relevant data is changed, propagate to all who use it or relate to it
-    if cached_data_changed? && table
-      user_table.dependent_visualizations.each(&:invalidate_cache)
+    if canonical? && (description_changed? || attributions_changed?)
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Can this check be moved to `InvalidationService`, at `invalidate`? Having one single method, `invalidate`, would be really nice.

File: spec/lib/carto/visualization_invalidation_service.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    @user = FactoryGirl.create(:carto_user)
+  end
+
+  after(:all) do
+    @user.destroy
+  end
+
+  shared_examples_for 'cache invalidator' do
+    it 'should invalidate all caches: redis/varnish vizjson and embed' do
+      CartoDB::Visualization::RedisVizjsonCache.any_instance.expects(:invalidate).with(@visualization.id)
```
Comment (juanignaciosl) [inline spec/lib/carto/visualization_invalidation_service.rb]: This lovely PR probably allows to put `.once` after those expectations :-)

File: spec/lib/carto/visualization_invalidation_service.rb [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    end
+  end
+
+  describe 'on update' do
+    before(:each) do
+      @map, @table, @table_visualization, @visualization = create_full_visualization(@user)
+      @service = Carto::VisualizationInvalidationService.new(@visualization)
+    end
+
+    # NOTE: Not destroying the visualizations in an after hook to avoid it triggering the wrogn expectations
```
Comment (juanignaciosl) [inline spec/lib/carto/visualization_invalidation_service.rb]: I think that you can `unstub` expectations, making the tests more accurate and enabling cleanup.

Typo: "wrogn".

File: app/models/carto/layer.rb (hunk 2 above)
Comment (javitonino) [inline app/models/carto/layer.rb]: I explained it a bit in the [commit message](https://github.com/CartoDB/cartodb/pull/12096/commits/8dfb35b9faf1b9543fb67fe3cb0ff18b77adcfb8), trying to get in the habit of writing some of those, since I find them useful while blaming.

Anyway, it has to do with association caching.
`maps.first`:
- If the association is cached, it works as expected
- If the association is not cached, it does a query with `LIMIT 1` and does not cache it for future use. Since the association is pretty much only accessed from here, it means it is never cached.
`maps[0]`
- If the association is cached, it works as expected
- If the association is not cached, it does a query without `LIMIT` and caches the association.

I changed this for the [`legend_spec`](https://github.com/CartoDB/cartodb/pull/12096/files#diff-7a85a0fdb681ae33bc1ee59e5c690f58L22) changes, since I noticed there that each call to `legend.layer.visualization` (used in the invalidation chain) was returning a different `map` instance, so it could call the invalidation hook multiple times.

File: app/models/carto/visualization.rb (hunk 3 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: - This change works as long as we are calling `save` (or `update_attributes`, etc) on any AR models. It has the non-obvious connotation that calling `Carto::Visualization.invalidate_later` does nothing if outside of a transaction (we could improve this by detecting if we are inside one and immediately invalidating in that case).
- For operations on old models, it uses a similar thing in `Visualization::Member` with an `after_commit` hook of the Sequel transaction (not at the model, but at the transaction itself, but is otherwise similar).
- For mixed operations (which we should have none of) where both ORMs are used, I think it will invalidate in both `after_commit` hooks ¯\_(ツ)_/¯

File: app/models/carto/visualization.rb (hunk 4 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: 1. 👍 

2. If there are multiple AR transactions, it will be called once per transaction. Most (all?) controllers only do one transaction. For mixed ORM transactions, this will probably invalidate twice, once per ORM.

3. Yes, more or less. We need to be careful with this.
  a. The old `Member` model never checked the return code of the invalidations. If it failed, it never noticed.
  b. The `Carto::Visualization` model was a bit better about this, and the transaction would be rollback in some case. This may be worse in some case (since there are some stuff outside the transaction, such as user db permissions).
  c. This PR also removes `save_named_map_or_rollback_privacy` which tried to rollback (including user db permissions) if the named map save failed. Not sure if that ever worked.

File: app/models/carto/visualization.rb (hunk 10 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: I know, but sadly I found no way of making it work. In `after_commit`, the `description_changed?` flag is already cleared, it always returns `false`.

💡 maybe `previous_changes` works here. Let me check...

File: lib/carto/visualization_invalidation_service.rb (hunk 7 above)
Comment (javitonino) [inline lib/carto/visualization_invalidation_service.rb]: I kept the original order, not sure about this one. I think it is irrelevant among the operations in this service, as the named map template generation is independent of vizjson/embed caches.

File: lib/carto/visualization_invalidation_service.rb (hunk 6 above)
Comment (javitonino) [inline lib/carto/visualization_invalidation_service.rb]: Ok, I did not like it that much to begin with. You get free suggestions by sending in shitty names to CR. Anyway, I'll get back to this after the previous comment, maybe we can get rid of this.

File: lib/carto/visualization_invalidation_service.rb (hunk 8 above)
Comment (javitonino) [inline lib/carto/visualization_invalidation_service.rb]: That's a nice idea now that we are refactoring. It could also simplify the code in `Visualization::Member`a bit, and hopefully in the future we will get an upset endpoint in maps API. 👍

File: lib/carto/visualization_invalidation_service.rb (hunk 9 above)
Comment (javitonino) [inline lib/carto/visualization_invalidation_service.rb]: No, we don't want to invalidate named maps for related maps. It was not done in the original code, and is also unnecessary (`description` and `attributions` are not presented in the named maps). We could add it and it would still be safe, but overkill.

File: spec/lib/carto/visualization_invalidation_service.rb (hunk 11 above)
Comment (javitonino) [inline spec/lib/carto/visualization_invalidation_service.rb]: Yes, but `expects` puts `once` by default :). We can add `once` jsut to be more explicit if you like (and I know you do) but it is not necessary.

File: spec/lib/carto/visualization_invalidation_service.rb (hunk 12 above)
Comment (javitonino) [inline spec/lib/carto/visualization_invalidation_service.rb]: Nope, I tested it. If you `unstub` expectations and they have not been called yet, they will not fail. If you add the `unstub` and remove the entire implementation of the invalidation service, the tests would pass 😞. So this is the best next thing I could figure out (I refuse to add `.times(2)` to the expectation to account for the destruction).
Typo 👍

File: app/models/carto/layer.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Meta: ok, let's give commit messages a try :-)

Nice to know that! ✨  Is that documented?

File: app/models/carto/visualization.rb (hunk 3 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Uhm!

> It has the non-obvious connotation that calling Carto::Visualization.invalidate_later does nothing if outside of a transaction 

I didn't realize that. I'd not only do what you suggest (detecting it and triggering invalidation) but also log it (we should probably do this always within a transaction).

👍  to other comments, thanks!

File: app/models/carto/visualization.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: 2. 👍 
3. Are there tests for `save_named_map_or_rollback_privacy`, or can we come up with them? That might be an issue (but nothing that we can't workaround).

File: app/models/carto/visualization.rb (hunk 10 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Oh, true. Don't make this more complicated if a simple change doesn't work, this is certainly good enough at this moment.

File: lib/carto/visualization_invalidation_service.rb (hunk 7 above)
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: :+1:

File: lib/carto/visualization_invalidation_service.rb (hunk 6 above)
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: xD

File: lib/carto/visualization_invalidation_service.rb (hunk 8 above)
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: ✨

File: lib/carto/visualization_invalidation_service.rb (hunk 9 above)
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: 👍 

I'd add a comment about it, though.

File: spec/lib/carto/visualization_invalidation_service.rb (hunk 11 above)
Comment (juanignaciosl) [inline spec/lib/carto/visualization_invalidation_service.rb]: xD Good to know, I _always_ put `once` xD. Following my explicitness obsession here is not necessary at all, don't worry :D

File: spec/lib/carto/visualization_invalidation_service.rb (hunk 12 above)
Comment (juanignaciosl) [inline spec/lib/carto/visualization_invalidation_service.rb]: I think that I've eventually done something similar to `times(2)` adding a comment about the destruction, or to `times(EXPECTED_CALLS + DESTRUCTION_CALLS)`. Anyway, it might be over-thinking. Up to you.

File: app/models/carto/layer.rb (hunk 2 above)
Comment (javitonino) [inline app/models/carto/layer.rb]: I haven't found a explicit site in the documentation where this is pointed out, but the thing is that accessing an association returns a [Relation](http://api.rubyonrails.org/v3.2.22/classes/ActiveRecord/Relation.html) object. In this object (which is collection-like), `first` is implemented as a query, while `[0]` force casting to an array, which executes the query, returns the results.

File: app/models/carto/visualization.rb (hunk 10 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: It works with something like this in the service:
`@visualization.canonical? && (@visualization.previous_changes.include?(:description) || @visualization.previous_changes.include?(:attributions))`

Let's try this approach, as it keeps the service caller dumb, and extracts everything to the service.

File: app/models/carto/visualization.rb (hunk 4 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: There are no tests that I can find (neither in the old model). I think this was already broken before this change, as `save_named_map` in `Member` [always returns `true`](https://github.com/CartoDB/cartodb/blame/12094-visualization_invalidation/app/models/visualization/member.rb#L639-L651) (as it enqueues the named map save in `after_commit`).  

We have been running without this for close to a year with no apparent issue. I say we torch it in the name of simplicity.

I've taken a second look at [`restore_previous_privacy`](https://github.com/CartoDB/cartodb/blame/12094-visualization_invalidation/app/models/visualization/member.rb#L784) which is what `save_named_map_or_rollback_privacy` was based on, and it only reverts the privacy of the visualization model, not syncing with the table model or the user DB, so it feels that this is worse than not reverting it.

File: app/models/carto/layer.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Being contributors is getting closer :-) Thanks!

File: app/models/carto/visualization.rb (hunk 10 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: ✨

File: app/models/carto/visualization.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: 👍  Let's stay with this, then.

File: app/models/carto/visualization.rb (hunk 3 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: Right now, it triggers a vague exception if you try calling `invalidate_after_commit` without a transaction:

```
NoMethodError: undefined method `last' for nil:NilClass
        from /home/javier/CartoDB/cartodb/vendor/bundle/ruby/2.2.0/gems/activerecord-3.2.22/lib/active_record/connection_adapters/abstract/database_statements.rb:243:in `add_transaction_record'
        from /home/javier/CartoDB/cartodb/vendor/bundle/ruby/2.2.0/gems/activerecord-3.2.22/lib/active_record/transactions.rb:298:in `add_to_transaction'
        from /home/javier/CartoDB/cartodb/app/models/carto/visualization.rb:539:in `invalidate_after_commit'
        from (irb):9
        from /home/javier/CartoDB/cartodb/vendor/bundle/ruby/2.2.0/gems/railties-3.2.22/lib/rails/commands/console.rb:47:in `start'
        from /home/javier/CartoDB/cartodb/vendor/bundle/ruby/2.2.0/gems/railties-3.2.22/lib/rails/commands/console.rb:8:in `start'
        from /home/javier/CartoDB/cartodb/vendor/bundle/ruby/2.2.0/gems/railties-3.2.22/lib/rails/commands.rb:41:in `<top (required)>'
        from script/rails:6:in `require'
        from script/rails:6:in `<main>'
```

This means I am pretty sure this is not happening right now in the code (as long as there is test coverage). I have mixed feelings about converting that hard failure into a logged warning plus doing what you intended. In the one hand, it is better than a cryptic error, in the other, I don't want this API to be used like this.

Do you have an strong opinion on this?

Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: In that case, I'd just trigger a better error.

**Comment by javitonino** [comment]
Tests were 🍏 after the tests, but I just merged master and it had some conflicts. I'd wait for this round of tests to pass before a second CR.

**Comment by javitonino** [comment]
After some hot merging action regarding permission refactor, this is ready to take the scene for a second time.

File: lib/carto/visualization_invalidation_service.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      elsif @visualization.data_layers.any?
+        named_maps_api.show ? named_maps_api.update : named_maps_api.create
+      end
+    end
+
+    def invalidate_affected_visualizations
+      # Invalidate the maps where this table is used. This is needed because the attributions of the datasets used
+      # in a map is used in the vizjson, and the description is shown in the public pages/embed.
+      # There is no need to invalidate the named maps, as nothing from the dataset is included there.
+      @visualization.user_table.dependent_visualizations.each do |affected_visualization|
```
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: As ` @visualization.user_table` can be `nil` I'd add a `return unless @visualization.user_table` at the beginning.

File: lib/carto/visualization_invalidation_service.rb (hunk 8 above)
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: Did you reject the change for any reason? It's minor, no need for it now.

File: lib/carto/visualization_invalidation_service.rb [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+    def invalidate_vizjson_from_varnish
+      CartoDB::Varnish.new.purge(".*#{@visualization.id}:vizjson")
+    end
+
+    def attributes_affecting_visualizations_changed?
+      @visualization.canonical? &&
+        (
+          @visualization.previous_changes.include?(:description) ||
+          @visualization.previous_changes.include?(:attributions)
```
Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: Minor: checking the intersection of two arrays might be better.

```ruby
ATTRIBUTES_AFFECTING_VISUALIZATIONS = [:description, :attributions].freeze

def attributes_affecting_visualizations_changed?
  @visualization.canonical? && !((@visualization.previous_changes & ATTRIBUTES_AFFECTING_VISUALIZATIONS).empty?)
end

File: lib/carto/visualization_invalidation_service.rb (hunk 8 above)
Comment (javitonino) [inline lib/carto/visualization_invalidation_service.rb]: Uh, I must have missed it while reviewing. As I've got to do a push anyway for other comments, I'll do it.

File: lib/carto/visualization_invalidation_service.rb (hunk 14 above)
Comment (javitonino) [inline lib/carto/visualization_invalidation_service.rb]: I tried this but it didn't work out too well. `previous_changes` is one of those indifferent hashes, so you ended up with `previous_changes.keys` plus making the constants out of strings, and after I really distrust that Rails keeps returning strings for indifferent access hashes anyway (Rails 4 migration has a bunch of issues with that).

Comment (juanignaciosl) [inline lib/carto/visualization_invalidation_service.rb]: Ok, keep this, then 👍

**Comment by javitonino** [comment]
Ok, did the last requested changes, and back to 🍏. I added some acceptance tips, feel free to add to it.

**Comment by juanignaciosl** [comment]
I added "This PR is totally related with caches, so please check request headers all the time :-)", I don't know if I should be more explicit.

**Comment by juanignaciosl** [comment]
@javitonino I haven't been able to test public map and embed cache because of a problem in staging (see link). Nevertheless, named maps update ok, and vizjson cache does as well, so this should be ok. Great work! 👏
